src/engine/source/logicexpr/interface/logicexpr/parser.hpp

```diff
@@ -0,0 +1,368 @@
+#ifndef _LOGICEXPR_PARSER_H
+#define _LOGICEXPR_PARSER_H
+
+#include <functional>
+#include <memory>
+#include <sstream>
+#include <stack>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#include <fmt/format.h>
+#include <utils/stringUtils.hpp>
+
+#include "tokenizer.hpp"
+
+/**
+ * @brief Namespace containing parsing logic expressions functionality
+ */
+namespace logicexpr::parser
+{
+
+namespace
+{
+// Helper to ensure correct syntax of expression in infix notation
+// Inspired by
+// https://stackoverflow.com/questions/29634992/shunting-yard-validate-expression
+struct syntaxChecker
+{
+
+    // Could be optimized by taking advantage of the fact that shunting-yard
+    // algorithm is already doing some comparisons, is as it is for modularity
+
+    // Two states:
+    // - 0: expect operand, unary operator or parenthesis
+    // - 1: expect binary operator
+    bool m_state;
+
+    bool expectedOperand() const { return m_state == false; }
+
+    void expectOperand() { m_state = false; }
+
+    bool expectedOperator() const { return m_state == true; }
+
+    void expectOperator() { m_state = true; }
+
+    syntaxChecker()
+        : m_state {false}
+    {
+    }
+
+    // Check if token satisfies the current state, otherwise throw exception
+    void operator()(const Token& token)
```

Comment:
nitpick - desing: Here the error messages are similar, I think you could do something like this:
```
    auto unexpectedTokenError = [&](const std::string& expected, const std::string& actual) {
        throw std::runtime_error(fmt::format("Unexpected {} '{}' at position '{}'", expected, actual, token->pos()));
    };

    if (token->isTerm()) {
        if (expectedOperator()) {
            unexpectedTokenError("token TERM", token->text());
        }
        expectOperator();
    } else if (token->isUnaryOperator()) {
        if (expectedOperator()) {
            unexpectedTokenError("unary operator", "NOT");
        }
        // Still wanting operand
    }
```